rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+# Reducing Primitive Creation Overhead for Dynamic Shapes
+
+## Motivation
+
+oneDNN API requires users to specify tensors dimensions to create a primitive,
+which makes the primitive dependent on the tensors's shapes. Primitive creation is
+an expensive operation, this is why oneDNN has a primitive cache to reduce the
+overhead. However, the primitive cache can only help if the tensors shapes do not
+change between workload iterations otherwise the primitives have to be created for
+the new shapes which incurs additional overhead. The goal of this RFC is to propose
+a solution to reduce that overhead.
+
+## Background
+
+There is a reason why the oneDNN API requires users to specify tensors dimensions
+to create a primitive. The library may use the information to generate a kernel
+specifically for the given shapes to get the best performance.
+
+However, the information may not always be used by the library meaning that the
+generated kernels can be used for different shapes, which makes requirement to
+always specify tensors dimensions is overly strict.
+
+There was an attempt in the past to relax the requirement by introducing so-called
+runtime dimensions support. The idea was to allow users specify a placeholder
+(`DNNL_RUNTIME_DIM_VAL`) instead of the actual dimensions. The general rule for
+the feature was that the more actual dimensions users provide the better
+performance they can get.
+
+It's obvious that in order to support runtime dimensions the kernels must not
+depend on the shapes. Or there should be different kernels that are specific to
+the amount of information provided. And the kernels should also demonstrate
+reasonably good performance otherwise, the creation overhead will be replaced by
+a suboptimal performance overhead, which will defeat the purpose of having the
+runtime dimensions feature.
+
+The runtime dimensions feature is not a silver bullet, this is why the RFC will
+consider another option to reduce the primitive creation overhead.
+
+## Proposal
+
+There are two options to reduce primitive creation overhead:
+1. Extend the runtime dimensions feature for the relevant primitives
+2. Introduce a cache for kernels
+
+### Option 1: Cache for Kernels (recommended)
+
+The idea is to introduce a cache for kernels (hereinafter *cache*). The cache will
+be global similar to the primitive cache.
+
+Pros:
+* Doesn't require users to modify their code to make the feature work
+* Eliminates the kernel creation overhead completely when it's possible
+* Doesn't require to expose kernel implementation details. For example, for
+the runtime dimensions feature we will have to document what dimensions can
+be runtime without introducing performance regressions
+
+Cons:
+* Primitive and primitive descriptor creation overhead will stay
```

Comment:
Thanks for the explanation. I can understand kernel cache benefits for GPU kernel generation, though it still depends on whether kernel can be re-used frequently.
For CPU kernels, looks the overhead reduction is uncertain.

Is there some initial evaluation on how much time can be reduced compare with current solution (with only primitive cache)?

Reply:
> For CPU kernels, looks the overhead reduction is uncertain.

Well, we can be certain that the kernel generation overhead will be eliminated. So the primitive creation will always take less time unless the kernel generation time is the same as the kernel cache lookup time, which is highly unlikely.

> Is there some initial evaluation on how much time can be reduced compare with current solution (with only primitive cache)?

Like I said above, we could just remove the kernel generation time to get an approximate estimation. It's hard (if even possible) to get an accurate common estimation because there are many factors that may affect the creation time.